### PR TITLE
pushsync, localstore: decouple push/pull indexes for tag increment

### PR DESCRIPTION
**Pull request by acud**
This PR aims to decrease the complexity that was added in #1828 by adding a `Tag` field to the `localstore` `pullIndex`. This allows decoupling of the logic of tag increments from the `pushIndex` and in addition, allows us to have the freedom of choosing whether to insert into `pushIndex` at all (this should not be done on all cases, for example when an anonymous upload is made).

As part of the changes needed, a backwards compatible `Tag` field has been appended to the `pullIndex` value definition. This field, due to the nature of the new tag increment logic is backwards compatible, thus there is not need for a database migration apart from just renaming the `pullIndex` using the new `shed` functionality of `RenameIndex`.

### CODE REVIEW

File: storage/localstore/mode_set.go [hunk 1]
```diff
@@ -201,48 +200,44 @@ func (db *DB) setSync(batch *leveldb.Batch, addr chunk.Address, mode chunk.ModeS
...
-					}
-				}
+	switch mode {
+	case chunk.ModeSetSyncPull:
+		// if we are setting a chunk for pullsync we expect it to be in the index
+		// if it has a tag - we increment it and set the index item to not contain the tag reference
+		// this prevents duplicate increments
+		i, err := db.pullIndex.Get(item)
+		if err != nil {
+			if err == leveldb.ErrNotFound {
```
Comment (acud) [inline storage/localstore/mode_set.go]: this peculiarity came from L238 on the left side

File: storage/localstore/mode_set_test.go [hunk 2]
```diff
@@ -65,6 +66,7 @@ func TestModeSetAccess(t *testing.T) {
 
 // TestModeSetSyncPull validates ModeSetSyncPull index values on the provided DB.
 func TestModeSetSyncPull(t *testing.T) {
+	t.Skip("for now")
```
Comment (acud) [inline storage/localstore/mode_set_test.go]: i think these tests can be removed now

**Comment by acud** [comment]
there is a data race in a stream test which is showing up here. I will fix this but I think it is orthogonal to this PR and that this can still be reviewed

File: storage/localstore/migration_test.go [hunk 3]
```diff
@@ -0,0 +1,306 @@
...
+			shouldNotRun = true // this should not be executed
+			return nil
+		}},
+	}
+
+	dir, err := ioutil.TempDir("", "localstore-test")
+	if err != nil {
+		t.Fatal(err)
+	}
+	cleanupFunc := func() { os.RemoveAll(dir) }
```
Comment (janos) [inline storage/localstore/migration_test.go]: Just a defer, would be ok, no need for cleanupFunc.

File: storage/localstore/migration.go [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+// with no errors in order to bring the localstore to the most up-to-date
+// schema definition
+func getMigrations(currentSchema, targetSchema string, allSchemeMigrations []migration) (migrations []migration, err error) {
+	foundCurrent := false
+	foundTarget := false
+	if currentSchema == DbSchemaCurrent {
+		return nil, nil
+	}
+	for i, v := range allSchemeMigrations {
+		v := v
```
Comment (janos) [inline storage/localstore/migration.go]: I do not think that `v := v` is needed.

File: storage/localstore/mode_set.go [hunk 5]
```diff
@@ -201,48 +201,65 @@ func (db *DB) setSync(batch *leveldb.Batch, addr chunk.Address, mode chunk.ModeS
...
-		case leveldb.ErrNotFound:
-			// the chunk is not accessed before
-		default:
+		if db.tags != nil && i.Tag != 0 {
+			t, err := db.tags.Get(i.Tag)
+
+			// increment if and only if tag is anonymous
+			if err == nil && t.Anonymous {
+				t.Inc(chunk.StateSent)
+				t.Inc(chunk.StateSynced)
```
Comment (zelig) [inline storage/localstore/mode_set.go]: we should not abuse the semantics of these states. 
We should not increment StateSynced

File: swarm.go [hunk 6]
```diff
@@ -497,7 +497,7 @@ func (s *Swarm) Stop() error {
 	}
 
 	if s.tags != nil {
-		err := s.stateStore.Put("tags", s.tags)
+		err := s.tags.Persist()
```
Comment (zelig) [inline swarm.go]: as we discussed this is not sufficient for correct tag persistance when crashes happen.

**Review by janos — COMMENTED** [review]
Only a few minor comments. I am not approving since it is still marked as in progress.

File: storage/localstore/mode_set.go [hunk 7]
```diff
@@ -201,48 +201,65 @@ func (db *DB) setSync(batch *leveldb.Batch, addr chunk.Address, mode chunk.ModeS
...
+				item.Tag = 0
```
Comment (zelig) [inline storage/localstore/mode_set.go:240]: This is super important, this line makes sure that syncing to several peers will not increment the same tag twice. 
Note that in push sync we delete synced chunks so we never increment twice. 
Please add a comment that explains this

**Review by zelig — CHANGES_REQUESTED** [review]
This PR is addressing 1) persistance, and 2) changing indexes. 
as for 1) i believe we may need more careful checkpoints as per roundtable discussion, no?
as for 2) synced state should not increment for pull sync only sent state.

**Comment by acud** [comment]
> This PR is addressing 1) persistance, and 2) changing indexes.
> as for 1) i believe we may need more careful checkpoints as per roundtable discussion, no?

TL;DR: our shutdown sequence does not promise that all goroutines related to push and pullsync are shut down before we persist tags. this has to be mended before we implement a dirty flag

Yes but as per my discussion with @janos it is not possible to implement the dirty flag right now, since we have no guarantees that once the dirty flag is unset (on persist), that no other goroutine is going to do a tag increment. example - on pushsync shutdown right now there is no guarantee that additional goroutines will _not_ do any other `Set` operations or tag increments. This will, in turn, very probably create a situation where the tags are constantly dirty on shutdown and thus on startup they will always be ignored and a new tag object created instead.

> as for 2) synced state should not increment for pull sync only sent state.

ok

File: chunk/tags.go [hunk 8]
```diff
@@ -98,21 +115,41 @@ func (ts *Tags) GetByAddress(address Address) (*Tag, error) {
...
+	if err != nil {
+		log.Error("had an error while persisting tags on delete", "err", err)
+	}
 }
 
-func (ts *Tags) Delete(k interface{}) {
-	ts.tags.Delete(k)
+// Persist calls the persistFunc with the current tags map
+// the caller is expected to hold the tags.mtx under write lock
+func (ts *Tags) Persist() error {
```
[the remark this answers is not in the document]
Comment (acud) [reply chunk/tags.go]: i'd leave it in since i don't want the call with the key scattered everywhere

File: storage/localstore/mode_set.go [hunk 9]
```diff
@@ -201,48 +201,72 @@ func (db *DB) setSync(batch *leveldb.Batch, addr chunk.Address, mode chunk.ModeS
...
+		if db.tags != nil && i.Tag != 0 {
+			t, err := db.tags.Get(i.Tag)
+			if err != nil {
+				// we cannot break or return here since the function needs to
+				// run to end from db.pushIndex.DeleteInBatch
+				log.Error("error getting tags on push sync set", "uid", i.Tag)
+			}
+
+			// setting a chunk for push sync assumes the tag is not anonymous
+			if err == nil && !t.Anonymous {
```
Comment (zelig) [inline storage/localstore/mode_set.go]: so how can it be anonymous if it is in the push sync index?

Comment (zelig) [inline storage/localstore/mode_set.go]:
```golang
else if t.Anonymous {
   log.Error("cannot and should not happen")
} else  {
  t.Inc(chunk.StateSynced)
}
```

**Review by zelig — COMMENTED** [review]
just one minor thing left

File: storage/localstore/mode_set.go (hunk 9 above)
Comment (acud) [reply storage/localstore/mode_set.go]: fixed

**Comment by janos** [comment]
The code cannot be compiled:

```
storage/localstore/migration.go:117:23: not enough arguments in call to db.pushIndex.Iterate
	have (func(shed.Item) (bool, error))
	want (shed.IndexIterFunc, *shed.IterateOptions)
storage/localstore/migration.go:120:5: continue is not in a loop
storage/localstore/migration.go:128:3: missing return at end of function
storage/localstore/migration.go:134:4: too many arguments to return
	have (nil, error)
	want (error)
```

File: storage/localstore/migration.go [hunk 10]
```diff
@@ -0,0 +1,142 @@
...
+
+// this function migrates Sanctuary schema to the Diwali schema
+func migrateSanctuary(db *DB) error {
+	// just rename the pull index
+	renamed, err := db.shed.RenameIndex("PO|BinID->Hash", "PO|BinID->Hash|Tag")
+	if err != nil {
+		return err
+	}
+	if !renamed {
+		return errors.New("pull index was not successfully renamed!")
```
Comment (janos) [inline storage/localstore/migration.go]: No need for exclamation sign in error message.

File: storage/localstore/migration.go [hunk 11]
```diff
@@ -0,0 +1,142 @@
...
+	}
+
+	ctr := 0
```
Comment (janos) [inline storage/localstore/migration.go]: No need for tracking number of operations in a batch, it is fine to write an empty batch and simplify code.

File: storage/localstore/migration.go [hunk 12]
```diff
@@ -0,0 +1,142 @@
...
+	batch := new(leveldb.Batch)
+	db.batchMu.Lock()
+	defer db.batchMu.Unlock()
+
+	// since pullIndex points to the Tag value, we should eliminate possible
+	// pushIndex leak due to items that were used by previous pull sync tag
+	// increment logic
+	if db.tags != nil {
```
Comment (janos) [inline storage/localstore/migration.go]: Avoid nesting with `db.tags == nil`.

File: storage/localstore/migration.go [hunk 13]
```diff
@@ -0,0 +1,142 @@
...
+			}
+			return false, nil
+		}, nil)
+		if err != nil {
+			return err
+		}
+	}
+
+	if ctr > 0 {
+		err = db.shed.WriteBatch(batch)
```
Comment (janos) [inline storage/localstore/migration.go]: This can be only `return db.shed.WriteBatch(batch)`.

File: storage/localstore/mode_put.go [hunk 14]
```diff
@@ -221,6 +221,13 @@ func (db *DB) putUpload(batch *leveldb.Batch, binIDs map[uint8]uint64, item shed
 	if exists {
 		return true, nil
 	}
+	anonymous := false
+	if db.tags != nil {
+		tag, err := db.tags.Get(item.Tag)
+		if err == nil {
```
Comment (janos) [inline storage/localstore/mode_put.go]: Not so correct to ignore error value. Anonymous tag should be set only if tag is not found, and error returned if it is not nil, even currently there is only tag not found error that can be returned here, but that is an implementation detail of tags that can change and allow incorrect logic here.

File: storage/localstore/mode_set.go [hunk 15]
```diff
@@ -191,7 +192,7 @@ func (db *DB) setSync(batch *leveldb.Batch, addr chunk.Address, mode chunk.ModeS
 		if err == leveldb.ErrNotFound {
 			// chunk is not found,
 			// no need to update gc index
-			// just delete from the push index
+			// just delete from the push indices
```
Comment (janos) [inline storage/localstore/mode_set.go]: No need for comment change.

File: storage/localstore/schema.go [hunk 16]
```diff
@@ -8,7 +24,7 @@ import (
 
 // The DB schema we want to use. The actual/current DB schema might differ
 // until migrations are run.
-const CurrentDbSchema = DbSchemaSanctuary
+var DbSchemaCurrent = DbSchemaDiwali
```
Comment (janos) [inline storage/localstore/schema.go:27]: I think that it would be more in compliance with naming to have a capitalised `B` in `Db` in all `DbSchema` variable names.

File: storage/localstore/migration_test.go [hunk 17]
```diff
@@ -0,0 +1,287 @@
...
+	}
+
+	err = db.Close()
+	if err != nil {
+		t.Error(err)
+	}
+}
+
+func TestManyMigrations(t *testing.T) {
+	DbSchemaCurrent = DbSchemaSanctuary
```
Comment (janos) [inline storage/localstore/migration_test.go:111]: DbSchemaCurrent should be restored after the test just it is done with allDbSchemaMigrations.

File: storage/localstore/schema.go [hunk 18]
```diff
@@ -24,6 +40,23 @@ const DbSchemaHalloween = "halloween"
 
 const DbSchemaSanctuary = "sanctuary"
 
+// the "diwali" migration simply renames the pullIndex in localstore
+const DbSchemaDiwali = "diwali"
+
+type migration struct {
+	name          string             //name of the schema
+	migrationFunc func(db *DB) error // the migration function that needs to be performed in order to get to the NEXT schema name
```
Comment (janos) [inline storage/localstore/schema.go]: Maybe to rename `migrationFunc` to just `func`, this type is already called migration.

File: storage/localstore/schema.go [hunk 19]
```diff
@@ -24,6 +40,23 @@ const DbSchemaHalloween = "halloween"
...
+}
+
+// allDbSchemaMigrations contains an ordered list of the database schemes, that is
+// in order to run data migrations in the correct sequence
+var allDbSchemaMigrations = []migration{
```
Comment (janos) [inline storage/localstore/schema.go]: Is `all` prefix needed? There is only one migrations slice. Can it be called just `migrations`?

File: storage/localstore/migration_test.go [hunk 20]
```diff
@@ -0,0 +1,287 @@
...
+
+	ran := false
+	shouldNotRun := false
+	allDbSchemaMigrations = []migration{
+		{name: DbSchemaSanctuary, migrationFunc: func(db *DB) error {
+			ran = true
+			return nil
+		}},
+		{name: DbSchemaDiwali, migrationFunc: func(db *DB) error {
+			shouldNotRun = true // this should not be executed
```
Comment (janos) [inline storage/localstore/migration_test.go]: This is in relation with my comment on `migrate` function. I do not know why last migration should not be executed.

File: storage/localstore/mode_set_test.go [hunk 21]
```diff
@@ -63,180 +64,341 @@ func TestModeSetAccess(t *testing.T) {
...
+	_, err = db.Put(context.Background(), chunk.ModePutUpload, ch)
+	if err != nil {
+		t.Fatal(err)
+	}
 
-					_, err = db.Put(context.Background(), chunk.ModePutUpload, chunks...)
-					if err != nil {
-						t.Fatal(err)
-					}
+	tag.Inc(chunk.StateStored) // so we don't get an error on tag.Status later on
```
Comment (janos) [inline storage/localstore/mode_set_test.go:86]: If this is needed in the test, should it be the responsibility of db.Put to increment StateStored if it finds the tag?

File: storage/localstore/mode_set_test.go [hunk 22]
```diff
@@ -63,180 +64,341 @@ func TestModeSetAccess(t *testing.T) {
...
-					defer cleanupFunc()
-
-					tag, err := db.tags.Create(mtc.name, int64(tc.count), mtc.anonymous)
-					if err != nil {
-						t.Fatal(err)
-					}
-					if tag.Uid != 0 {
-						t.Fatal("expected mock tag uid")
-					}
+// here we try to set a normal tag (that should be handled by pushsync)
```
Comment (janos) [inline storage/localstore/mode_set_test.go:68]: It looks to me that there some code deduplication can be done in new tests. The previous test had too many variables for subtests, but maybe some test helpers can reduce the duplication in tests. Just an idea.

File: storage/localstore/mode_set.go [hunk 23]
```diff
@@ -201,48 +202,74 @@ func (db *DB) setSync(batch *leveldb.Batch, addr chunk.Address, mode chunk.ModeS
...
-				} else {
-					// not anonymous - push and pull should be used
-					switch mode {
-					case chunk.ModeSetSyncPull:
-						moveToGc = false
-					case chunk.ModeSetSyncPush:
-						tag.Inc(chunk.StateSynced)
-					}
-				}
+	switch mode {
```
Comment (janos) [inline storage/localstore/mode_set.go:209]: It would be good to update setSync comment with some notion on anonymous tags, so that the exported api can be documented.

**Review by janos — CHANGES_REQUESTED** [review]

File: storage/localstore/migration.go [hunk 24]
```diff
@@ -0,0 +1,142 @@
...
+
+	if migrations != nil {
+		log.Info("need to run data migrations on localstore", "numMigrations", len(migrations), "schemaName", schemaName)
+		for i := 0; i < len(migrations)-1; i++ {
+			err := migrations[i].migrationFunc(db)
+			if err != nil {
+				return err
+			}
+			if i != len(migrations)-1 {
+				err = db.schemaName.Put(migrations[i+1].name) // put the name of the next schema
```
[the remark this answers is not in the document]
Comment (acud) [reply storage/localstore/migration.go]: > Why are we never running migrations for the last element? But also setting the schema of the last migration.

if you would look at the definition of the `migration` struct:
```go
type migration struct {
	name          string             //name of the schema
	migrationFunc func(db *DB) error // the migration function that needs to be performed in order to get to the NEXT schema name
}
```

maybe this is not intuitive enough. i can change it that the migrationFunc will lead to the current schema name, not the next.

> When any migration is done, its schema should be stored in schemaName field, so that if any intermediate migration fails, the migration can be restarted after the last successful one.

that is actually what happens. but yeah i guess this code is a bit overly convoluted. i will refactor this

File: storage/localstore/migration.go (hunk 12 above)
Comment (acud) [reply storage/localstore/migration.go]: actually, this should never happen. tags object should never be `nil`

File: storage/localstore/schema.go (hunk 16 above)
Comment (acud) [reply storage/localstore/schema.go:27]: compliance with which other convention? you originally added it with `Db` not `DB` :)

File: storage/localstore/mode_set_test.go (hunk 21 above)
Comment (acud) [reply storage/localstore/mode_set_test.go:86]: no, StateStored is incremented in `hasherstore`..... remove this line and you will understand the error that is invoked. this must be called in the test because of the `tag.Status` logic. please review that function to understand why this is called

File: storage/localstore/schema.go (hunk 16 above)
Comment (acud) [reply storage/localstore/schema.go:27]: as an example we also usually use `Uid` not `UID`, `Ruid` as opposed to `RUID`, etc. if anything, we should first conclude on how we capitalize acronyms in the first place.
personally I am not such a big proponent of CAPS in the codebase. I find lowercase easier on the eyes

File: storage/localstore/mode_set_test.go (hunk 22 above)
Comment (acud) [reply storage/localstore/mode_set_test.go:68]: removed some of the test code with the `chunk/testing#CheckTag` helper function. not sure if there's anything else to remove here. i find this manageable to read

File: storage/localstore/mode_set.go (hunk 23 above)
Comment (acud) [reply storage/localstore/mode_set.go:209]: added comment to the enclosing function. let me know if it's enough

File: storage/localstore/schema.go (hunk 16 above)
Comment (janos) [reply storage/localstore/schema.go:27]: I may be wrong, but I do not think that I added schema name variables. I used DB in shed. 

The convention that is usually noted is this one https://github.com/golang/go/wiki/CodeReviewComments#initialisms on initialisms. I also think that `Uid` should be `UID` and so on, but I missed to review these pull requests.

The uppercaps argument does not stand, as it is about "same" caps, `DB` can be `db` at the start of unexported value or on its own.

In any case, we have so much inconsistencies in the code that this does not matter at all.

File: storage/localstore/mode_set_test.go (hunk 21 above)
Comment (janos) [reply storage/localstore/mode_set_test.go:86]: Thanks for the explanation.

**Review by janos — APPROVED** [review]
Thanks for addressing my comments. LGTM.

File: storage/localstore/schema.go (hunk 16 above)
Comment (acud) [reply storage/localstore/schema.go:27]: cool, i agree with you, let's stick to the golang recommendations for next PRs. we can actually do this change as a separate PR and rename across the codebase for consistency

**Review by zelig — APPROVED** [review]

File: storage/localstore/migration_test.go [hunk 25]
```diff
@@ -0,0 +1,382 @@
...
+// to the most recent schema.
+func TestMigrateSanctuaryFixture(t *testing.T) {
+
+	tmpdir, err := ioutil.TempDir("", "localstore-test")
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer os.RemoveAll(tmpdir)
+
+	cdir, err := os.Getwd()
```
Comment (janos) [inline storage/localstore/migration_test.go]: This is not needed. Tests are run always in current directory. Relative path for `dir` is fine.

File: storage/localstore/migration_test.go [hunk 26]
```diff
@@ -0,0 +1,381 @@
...
+	if err != nil {
+		return err
+	}
+	defer in.Close()
+	out, err := os.Create(dst)
+	if err != nil {
+		return err
+	}
+	defer func() {
+		err = out.Close()
```
Comment (janos) [inline storage/localstore/migration_test.go]: This error will shadow any returned one. It is possible that Copy or Sync return not nil error and that the err is set to nil, by the Close in this defer.

File: storage/localstore/migration_test.go [hunk 27]
```diff
@@ -0,0 +1,381 @@
...
+	}()
+	if _, err = io.Copy(out, in); err != nil {
+		return err
+	}
+	err = out.Sync()
```
Comment (janos) [inline storage/localstore/migration_test.go]: This error is never returned.
